packages/gatsby-plugin-sharp/src/gatsby-node.js

```diff
@@ -1,10 +1,96 @@
-const { setBoundActionCreators, setPluginOptions } = require(`./index`)
+const {
+  setBoundActionCreators,
+  setPluginOptions,
+  queue: jobQueue,
+  reportError,
+} = require(`./index`)
+const { scheduleJob } = require(`./scheduler`)
 
-exports.onPreInit = ({ actions }, pluginOptions) => {
+const getQueueFromCache = async cache => {
+  const rawQueue = await cache.get(`queue`)
+
+  if (!rawQueue) {
+    return new Map()
+  }
+
+  return new Map(rawQueue)
+}
+
+const saveQueueToCache = async (cache, queue) => {
+  const cachedQueue = await getQueueFromCache(cache)
+
+  // merge both queues
+  for (const [key, job] of cachedQueue) {
+    if (!queue.has(key)) {
+      queue.set(key, job)
+    }
+  }
+
+  // JSON.stringify doesn't work on an Map so we need to convert it to an array
+  return cache.set(`queue`, Array.from(queue))
```

Comment:
This is only thing that could potentially be blocker.

Just to get it out of the way - I suggested using `cache` initially, so this is my fault.

Since we started looking into cache issues, I started paying more attention to potential problems with it. Problem with using `cache` instead of redux state here would be that we potentially can have inconsistent state. 

Processing queue is linked to query running (not exclusively but major usage of it is). State that triggers query running is stored in redux state (there are other triggers, but things we persist is stored there). Scenario when processing queue fails to persist (to `cache`), but redux state persist correctly can lead to situations when images that are supposed to be created during production build, aren't, and build completes successfully (no errors thrown).

Long story short - we can't store related chunks of data in different persistence engines. We need to use redux state here. 

Luckily we have easy way to do that through [setPluginStatus](https://www.gatsbyjs.org/docs/actions/#setPluginStatus) action:
```
exports.whateverAPI = async ({ actions }) => {
  actions.setPluginStatus({ queue: Array.from(queue) })
}
``` 

Reading it is not as nice, but still OK-ish:
```js
let getPluginState = null
exports.onPreBootstrap = ({ store }) => {
  getPluginState = () => {
    const pluginStatus = store.getState().status.plugins[`gatsby-plugin-sharp`]
    if (pluginStatus && pluginStatus.queue) {
      return new Map(pluginStatus.queue)
    }
    return new Map()
  }
}
```
(ideally we add helper function for this to gatsby, but I would suggest using this to be backward compatible as there aren't any breaking changes here)

So all this can potentially increase problems with persisting redux state (higher memory usage / long writes), which is not great :/ Some issues for reference - https://github.com/gatsbyjs/gatsby/issues/8125 https://github.com/gatsbyjs/gatsby/issues/6611 https://github.com/gatsbyjs/gatsby/issues/6656

So I hate to even suggest this given how great feature this is - but should we give a way to opt-out of it? (if other folks agree with my essay that is - I would very happy to be proven wrong about inconsistency issue - it happened before)

Reply:
@pieh thanks a lot! I'll see what I can do with it. I'll add a new pluginOption `lazyImageGenaration` (if you have a better please tell me). Which is set by default on true.

It's sad to hear that redux state persisting might cause memory issues. We maybe should try to move to something else or be smarter about our store, like don't load it upfront unless it's asked for and clear objects when they aren't accessed anymore

Reply:
I did the changes, unsure if this is what you meant 😛 @pieh

Reply:
Memory issues mostly are related to saving state to file - and not saving part but serializing ( https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/redux/index.js#L91 ).

Just the way `stringify` (either `JSON.stringify` or one from `json-stringify-safe`) works, is that it creates many intermediatte strings which add up in memory and not being garbage collected quick enough by Node's V8 engine.

We tried some streaming alternatives but this wasn't really working for us (at least our implementation of using it). See https://github.com/gatsbyjs/gatsby/pull/9896 for revert of it and linked comment in PR description that summarise issue we had with it.